### PR TITLE
prepare for vsts microbuild pipeline

### PR DESCRIPTION
**Pull request by rohit21agrawal**
This PR will not have any impact on the current TeamCity build pipeline, but adds some extra targets for loc/signing in the VSTS MicroBuild environment.

CC: @rrelyea @emgarten @nkolev92 @mishra14 @alpaix

### CODE REVIEW

File: build/common.targets [hunk 1]
```diff
@@ -90,6 +96,95 @@
...
+      Targets="_GetLocalizedDllsForSigning"
+      Properties="IsVsixBuild=true;
+                  BuildProjectReferences=false;">
+      <Output
+        TaskParameter="TargetOutputs"
+        ItemName="DllsToSign"/>
+    </MsBuild>
+  </Target>
+
+  <Target Name="_GetLocalizedDllsForSigning" Returns="@(LocalizedDllsToSign)">
```
Comment (nkolev92) [inline build/common.targets:126]: Discussed offline with Rohit, this target will pick-up all the dll included in vsix and try to localize them even if they have no resources.

File: src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj [hunk 2]
```diff
@@ -358,13 +352,15 @@
       <AllVsixAssemblies Include="@(VSIXSourceItem->'%(FileName)%(Extension)')" Condition=" '%(Extension)' == '.dll' ">
         <SourceFile>%(VSIXSourceItem.Identity)</SourceFile>
       </AllVsixAssemblies>
+      <ScriptFiles Include="$(OutputPath)Scripts\*"/>
```
Comment (rohit21agrawal) [inline src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj]: get rid of this, not needed anymore

File: src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj [hunk 3]
```diff
@@ -358,13 +352,15 @@
...
     </ItemGroup>
     <ItemGroup>
       <SuppressPackaging Include="@(AllVsixAssemblies)" Exclude="@(IncludeVsixFilesList)" />
       <MissingRequiredAssemblies Include="@(IncludeVsixFilesList)" Exclude="@(AllVsixAssemblies)" />
       <NotListedDependencies Include="@(SuppressPackaging)" Exclude="@(IgnoreVsixFilesList)" />
       <MissingIgnoredAssemblies Include="@(IgnoreVsixFilesList)" Exclude="@(AllVsixAssemblies)" />
       <VSIXSourceItem Remove="@(SuppressPackaging->'%(SourceFile)')" />
+      <VSIXSourceItem Remove="@(ScriptFiles)" />
```
Comment (rohit21agrawal) [inline src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj]: no need.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.nuspec [hunk 4]
```diff
@@ -1,18 +1,18 @@
-<?xml version="1.0"?>
+﻿<?xml version="1.0"?>
 <package>
   <metadata>
     <id>NuGet.Build.Tasks.Pack</id>
     <version>$version$</version>
     <authors>Microsoft</authors>
     <owners>Microsoft</owners>
-    <description>NuGet 3 pack for dotnet CLI</description>
+    <description>NuGet  pack for dotnet CLI</description>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.nuspec]: nit: extra space

File: test/NuGet.Core.FuncTests/NuGet.Core.FuncTest/DownloadTimeoutStreamTests.cs [hunk 5]
```diff
@@ -83,7 +83,7 @@ public void DownloadTimeoutStream_RejectsNullStream()
             await VerifyTimeoutOnReadAsync(stream => Task.FromResult(ReadStream(stream)));
         }
         
-        [Fact]
+        //[Fact]
         public async Task DownloadTimeoutStream_FailureSync()
```
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/NuGet.Core.FuncTest/DownloadTimeoutStreamTests.cs]: I assume we don't want this disabled.

File: build/build.proj [hunk 6]
```diff
@@ -242,12 +242,14 @@
...
   -->
-  <Target Name="BuildVS15VSIX" AfterTargets= "BuildVS15" Condition=" '$(IsXPlat)' != 'true' ">
+  <Target Name="BuildVS15VSIX" AfterTargets= "BuildVS15" Condition=" '$(IsXPlat)' != 'true' AND '$(BuildRTM)' != 'true'">
     <Message Text="Building the VSIX for VS15" Importance="high" />
 
     <MSBuild Projects="$(VSIXProject)"
              Targets="Build"
              Properties="$(CommonMSBuildProperties);
+                         BuildProjectReferences=false;
+                         IsVsixBuild=true;
```
Comment (nkolev92) [inline build/build.proj:212]: Do we want to have the same thing for VS14 for the sake of consistency?

File: build/loc.proj [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      BuildRTM=$(BuildRTM);
+    </CommonMSBuildProperties>
+    <OutDir>$(ArtifactsDirectory)</OutDir>
+    <IntermediateOutputPath>$(ArtifactsDirectory)</IntermediateOutputPath>
+  </PropertyGroup>
+  
+  <!-- All projects in the repository -->
+  <ItemGroup> 
+      <SolutionProjects Include="$(RepositoryRootDirectory)src\*\*\*.csproj" />
+      <SolutionProjectsWithoutVSIX Include="@(SolutionProjects)"
```
Comment (nkolev92) [inline build/loc.proj]: We have this defined in 3 places, maybe extract the common stuff from sign/loc/build.proj somewhere?

**Review by nkolev92 — APPROVED** [review]
LGTM

Address the comments and then 🚢

File: build/common.project.props [hunk 8]
```diff
@@ -9,6 +9,7 @@
   <!-- Common -->
   <PropertyGroup>
     <RepositoryRootDirectory>$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), 'README.md'))\</RepositoryRootDirectory>
+    <LocalizationRootDirectory>$(RepositoryRootDirectory)submodules\NuGet.Build.Localization\localize</LocalizationRootDirectory>
```
Comment (alpaix) [inline build/common.project.props]: The value here can be defined as `$(NuGetBuildLocalizationRepository)localize`.

File: build/common.targets [hunk 9]
```diff
@@ -55,6 +53,14 @@
     <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
   </PropertyGroup>
 
+  <!-- These are needed because currently a VSIX project cannot reference a net core project that multi targets
+  This is a hack for building VS14 VSIX and -->
```
Comment (alpaix) [inline build/common.targets]: and ?

File: build/common.targets [hunk 10]
```diff
@@ -90,6 +96,95 @@
 
   <!--
     ============================================================
+    GatherSigningInputs - gathers the list of DLLs that need to
```
Comment (alpaix) [inline build/common.targets]: `GetSigningInputs`

File: src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop/NuGet.SolutionRestoreManager.Interop.csproj [hunk 11]
```diff
@@ -6,11 +6,19 @@
...
+  <ItemGroup>
+    <None Update="install.ps1">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
   <PropertyGroup>
     <NuspecFile>NuGet.SolutionRestoreManager.Interop.nuspec</NuspecFile>
     <NuspecProperties>version=$(Version);configuration=$(Configuration)</NuspecProperties>
+    <NuspecBasePath>$(OutputPath)</NuspecBasePath>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop/NuGet.SolutionRestoreManager.Interop.csproj:21]: 👍

File: src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj [hunk 12]
```diff
@@ -3,10 +3,12 @@
 
   <PropertyGroup>
     <Description>Complete commands common to command-line and GUI NuGet clients</Description>
-    <TargetFrameworks>netstandard1.3;net45</TargetFrameworks>
+    <TargetFrameworks Condition="'$(IsVsixBuild)' != 'true'">netstandard1.3;net45</TargetFrameworks>
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj:6]: Duplicated many times. Include in `common.project.props` instead?

**Review by alpaix — APPROVED** [review]
:sparkles: Address minor feedback notes.

File: build/build.proj (hunk 6 above)
Comment (rohit21agrawal) [reply build/build.proj:212]: no, i am keeping the VS14 as is, and VSTS isn't going to be building VS14 most likely.

File: src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj (hunk 12 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj:6]: i talked to @emgarten abut this. it turns out that the netcore project system starts treating everything relative to the file which has the $(TargetFramework) property, and that's why we can't move it to common.project.props. Justin, please correct me if i am wrong.

File: build/loc.proj [hunk 13]
```diff
@@ -68,6 +61,18 @@
...
+    <ItemGroup>
+      <!-- SupportedCultures is defined in MicroBuild.Plugins.Localization.targets which is installed as a plugin
+      in VSTS MicroBuild -->
+      <EulaFiles Include="$(NuGetBuildLocalizationRepository)eulas\%(SupportedCultures.Identity)\eula.rtf"
+          Condition="Exists('$(NuGetBuildLocalizationRepository)eulas\%(SupportedCultures.Identity)\eula.rtf')">
+          <DestinationDir>$(ArtifactsDirectory)eulas\%(SupportedCultures.FNCulture)\eula.rtf</DestinationDir>
+      </EulaFiles>
+    </ItemGroup>
+    <Copy SourceFiles="@(EulaFiles->'%(Identity)')" DestinationFiles="@(EulaFiles->'%(DestinationDir)')"/>
+  </Target>
```
Comment (nkolev92) [inline build/loc.proj:74]: Cool 🥇

**Review by emgarten — APPROVED** [review]

**Pull request merged**